### PR TITLE
Simplify ASR.asdl

### PR DESCRIPTION
**Pull request by khushi-411**
Fixes: #1900

#### TODO
- [x] add doccumentation

### CODE REVIEW

**Comment by certik** [comment]
I think this looks good so far.

**Comment by certik** [comment]
We can move all the comments into `doc/src/asr/asr.md`, and just ensure they are all there. Then we can merge this PR.

After that, we can iteratively improve the documentation to move things around (split into individual nodes) and improve.

File: src/libasr/ASR.asdl [hunk 1]
```diff
@@ -1,177 +1,25 @@
--- Abstract Semantic Representation (ASR) definition
-
```
Comment (certik) [inline src/libasr/ASR.asdl:2]: I would keep this line and underneath I would put a link to the full documentation (`doc/src/asr/asr.md`).

File: doc/src/asr/asr.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+> ***Note***: symbol_table contains `identifier` -> `symbol` mappings
+
+In the above example, `symbol_table`, `identifier`, `stmt`, `bool`, etc are types.
+
+#### Constructors
+The **constructors** names must begin with an upper case.
+In above example has three constructors, `Program`, `Module`, and `Function`,
+where the Program constructor has four fields whose values are of type symbol_table,
+identifier, identifier *, and stmt *. These are basically, subtrees.
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:56]:
```suggestion
where the Program constructor has four fields whose values are of type `symbol_table`,
`identifier`, `identifier *`, and `stmt *`. These are basically, subtrees.
```

File: doc/src/asr/asr.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+## Symbol Table
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:58]:
```suggestion
## Symbol type
```
Should it be type?

File: doc/src/asr/asr.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+Each symbol has either `symtab` (local symbol table) or `parent_symtab`
+(where this symbol is stored). One can get to parent_symtab via symtab, so
+only one is present.
+
+Each symbol has a `name` for easy lookup of the name of the symbol when only
+having a pointer to it.
+
+Each symbol has a `name` for easy lookup of the name of the symbol when only
+having a pointer to it.
+
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:69]:
```suggestion
```
Repeated.

File: doc/src/asr/asr.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+abi=Source means the symbol's implementation is included (full ASR),
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:70]:
```suggestion
`abi=Source` means the symbol's implementation is included (full ASR),
```

File: doc/src/asr/asr.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+otherwise it is external (interface ASR, such as procedure interface).
+
+SubroutineCall/FunctionCall store the actual final resolved subroutine or
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:73]:
```suggestion
`SubroutineCall`/`FunctionCall` store the actual final resolved subroutine or
```

File: doc/src/asr/asr.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+function (`name` member). They also store the original symbol
+(`original_name`), which can be one of: null, GenericProcedure or
+ExternalSymbol.
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:76]:
```suggestion
(`original_name`), which can be one of: `null`, `GenericProcedure` or
`ExternalSymbol`.
```

File: doc/src/asr/asr.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+- **abi=Interactive**: the symbol's implementation has been provided by the
+previous REPL execution (e.g., if LLVM backend is used for the interactive
+mode, the previous execution generated machine code for this symbol's
+implementation that was loaded into memory). Note: this option might be
+converted/eliminated to just use LFortran ABI in the future.
+- **abi=Intrinsic**: the symbol's implementation is implicitly provided by the
+language itself as an intrinsic function. That means the backend is free to
+implement it in any way it wants. The function does not have a body, it is
+just an interface.
+
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:139]:
```
abi                   -- External     ABI
    = Source          --   No         Unspecified
    | LFortranModule  --   Yes        LFortran
    | GFortranModule  --   Yes        GFortran
    | BindC           --   Yes        C
    | BindPython      --   Yes        Python
    | BindJS          --   Yes        Javascript
    | Interactive     --   Yes        Unspecified
    | Intrinsic       --   Yes        Unspecified
```
Should we keep the above table?

File: doc/src/asr/asr.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    Maybe it's easier to add an enumeration here:
+    - Descriptor -> Pointer
+    - Pointer -> Descriptor
+    - CompileTimeFixedSizeArray -> Pointer
+    - CompileTimeFixedSizeArray -> Descriptor
+    - Descriptor -> NumPy
+    - NumPy -> Descriptor
+    - ISODescriptor -> Descriptor
+    - Descriptor -> ISODescriptor
+3. **ttype**
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:158]:
```suggestion
### Ttype nodes
```

File: doc/src/asr/asr.md [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+integer, so we also use kind=4 for the default logical.)
+
+### Stmt nodes
+1. **ExplicitDeallocate**: It deallocates if allocated otherwise throws a runtime error.
+2. **ImplicitDeallocate**: It  deallocates if allocated otherwise does nothing.
+3. **GoTo**: It points to a GoToTarget with the corresponding target_id within
+the same procedure. We currently use `int` IDs to link GoTo with
+GoToTarget to avoid issues with serialization.
+4. **GoToTarget**: An empty statement, a target of zero or more GoTo statements
+the `id` is only unique within a procedure.
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:188]:
```
string_format_kind
    = FormatFortran        -- "(f8.3,i4.2)", a, b
    | FormatC              -- "%f: %d", a, b
    | FormatPythonPercent  -- "%f: %d" % (a, b)
    | FormatPythonFString  -- f"{a}: {b}"
    | FormatPythonFormat   -- "{}: {}".format(a, b)
```
I think we need to keep the above documentation as well

File: doc/src/asr/asr.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    - Complex kinds: 4 (c32), 8 (c64)
+    - Character kinds: 1 (utf8 string)
+    - Logical kinds: 1, 2, 4: (boolean represented by 1, 2, 4 bytes; the default
+kind is 4, just like the default integer kind, consistent with Python
+and Fortran: in Python "Booleans in Python are implemented as a subclass
+of integers", in Fortran the "default logical kind has the same storage
+size as the default integer"; we currently use kind=4 as default
+integer, so we also use kind=4 for the default logical.)
+
+### Stmt nodes
```
Comment (Thirumalai-Shaktivel) [inline doc/src/asr/asr.md:181]: Minor, I think we need to move stmt nodes above the expr nodes

File: doc/src/asr/asr.md (hunk 7 above)
Comment (Thirumalai-Shaktivel) [reply doc/src/asr/asr.md:76]: Let's highlight all the ASR nodes in the documentation

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
Rest looks good, thank you!

File: doc/src/asr/asr.md (hunk 8 above)
Comment (khushi-411) [reply doc/src/asr/asr.md:139]: Sure! updated

File: src/libasr/ASR.asdl [hunk 12]
```diff
@@ -417,97 +203,35 @@ ttype
...
 require_instantiation = Require(identifier name, identifier* args)
-
-string_format_kind
-    = FormatFortran        -- "(f8.3,i4.2)", a, b
-    | FormatC              -- "%f: %d", a, b
-    | FormatPythonPercent  -- "%f: %d" % (a, b)
-    | FormatPythonFString  -- f"{a}: {b}"
-    | FormatPythonFormat   -- "{}: {}".format(a, b)
+-- TODO: prefix the enumerators here, improve the names
```
Comment (certik) [inline src/libasr/ASR.asdl:227]: Remove this one

File: src/libasr/ASR.asdl [hunk 13]
```diff
@@ -235,25 +74,19 @@ stmt
     | DictInsert(expr a, expr key, expr value)
     | Expr(expr expression)
 
-
 expr
     = IfExp(expr test, expr body, expr orelse, ttype type, expr? value)
         -- Such as: (x, y+z), (3.0, 2.0) generally not known at compile time
```
Comment (certik) [inline src/libasr/ASR.asdl:79]: Remove this one

File: doc/src/asr/asr.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
+# Abstract Syntax Representation (ASR)
```
Comment (certik) [inline doc/src/asr/asr.md:1]:
```suggestion
# Abstract Semantic Representation (ASR)
```

**Comment by certik** [comment]
I think this looks great. Fix up the conflicts and the few things I noticed, then we can merge and improve upon it later.

**Comment by khushi-411** [comment]
Done! Thank you for all your prompt reviews in this PR :-)

File: src/libasr/ASR.asdl (hunk 13 above)
Comment (certik) [reply src/libasr/ASR.asdl:79]: This is not resolved yet.

**Comment by certik** [comment]
Thanks. One last thing: https://github.com/lfortran/lfortran/pull/3063#discussion_r1448137930.

**Review by certik — APPROVED** [review]
I think it looks good now. Now all comments were ported, but most of them were, so I think this is good enough to merge. Thanks!

**Pull request merged**